beetsplug/audible.py

```diff
@@ -104,11 +248,117 @@ def __init__(self):
+            else:
+                matches = sorted_tracks
+            return matches
+
+    def attempt_match_starting_numbers(self, items: List[Item], album: AlbumInfo) -> Optional[List[Item]]:
+        """Order tracks based on a starting number in the track name."""
+        affixes = find_regular_affixes([c.title for c in items])
+        stripped_titles = [strip_affixes(i.title, affixes) for i in items]
+
+        starting_numbers = [check_starts_with_number(s) for s in stripped_titles]
+        if all([s is not None for s in starting_numbers]) and is_continuous_number_series(sorted(starting_numbers)):
+            items_with_numbers = list(zip(starting_numbers, items))
+            matches = sorted(items_with_numbers, key=lambda i: i[0])
+            matches = [i[1] for i in matches]
+            return matches
+
+    def attempt_match_natsort(self, items: List[Item], album: AlbumInfo) -> Optional[List[Item]]:
+        """Use a natural sort on the input tracks to order them like a person would i.e. 10 is after 9, not 2."""
+        affixes = find_regular_affixes([c.title for c in items])
+        stripped_titles = [strip_affixes(i.title, affixes) for i in items]
+        average_title_change = calculate_average_levenshtein_difference(stripped_titles)
+        # magic number here, it's a judgement call
+        if max(average_title_change) < 4:
+            # can't assume that the tracks actually match even when there are the same number of items, since lengths
+            # can be different e.g. an even split into n parts that aren't necessarily chapter-based so just natsort
+            matches = natsorted(items, key=lambda t: t.title)
+            return matches
+
+    def attempt_match_chapter_levenshtein(self, items: List[Item], album: AlbumInfo) -> Optional[List[Item]]:
+        """For every chapter in the input album, calculate the Levenshtein difference between every item in the online
+        album and match each input track to the closest online track."""
+        # Warning, this method is rather messy, and it's easy for this to go wrong.
+        # This should be used as a last resort
+        affixes = find_regular_affixes([c.title for c in items])
+
+        all_remote_chapters: List = deepcopy(album.tracks)
+        matches = []
+        for chapter in items:
+            # TODO: need a string distance algorithm that penalises number replacements more
+            best_matches = list(
+                sorted(
+                    all_remote_chapters,
+                    key=lambda c: specialised_levenshtein(chapter.title, c.title, affixes),
+                )
+            )
+            try:
+                best_match = best_matches[0]
+            except IndexError:
+                return None
+            matches.append(best_match)
+            all_remote_chapters.remove(best_match)
+        return matches
+
+    def attempt_match_single_item(self, items: List[Item], album: AlbumInfo) -> Optional[List[Item]]:
```

Comment:
I like this approach when there is only one item.

There are some fairly common cases where the book has a few files (e.g, part 1 and part 2). Would those situations fall back to the other approaches further down the list, and how well do they work for multi-part books?

Reply:
Yes, this is the idea. Each method is tried after the previous methods have failed. The given example configuration is, in my opinion, the best way to order them, though it can change according to what people want and what files their library has.

In this case, if there is more than one file, then the other methods will be tried. If the files have numbering in their metadata with no missing numbers, this order will be used. If they can be sorted naturally (the part 1 and part 2 case you mentioned) then that will be tried next. 

I've found that this works extremely well for almost every type of book. There are a few options that I would like to make configurable, mainly the various magic numbers that I've marked, but overall this is a very comprehensive system that I feel can be easily extended if more algorithms become apparent.